### PR TITLE
Toanzian/studio merge

### PR DESCRIPTION
**Pull request by Fignewton5**
- Added card editor components
  - Monaco JSON editor
  - Live card preview
  - Card template entities window (WIP)
  - Output window that displays card preview actions

**NOTE**: The card editor currently uses a card reducer that controls the state for the current card being edited. I assume this will change to probably just loading the card state from the selected document out of state.editor.documents, but it's just temporary. The structure of the card state that I'm using is also a WIP, as it will probably need to hold a path key as well that points to the card's location in the file system.

### CODE REVIEW

File: client-v2/src/data/reducer/card.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
+import {
```
Comment (eanders-MS) [inline client-v2/src/data/reducer/card.js]: This should appear below the copyright notice

File: client-v2/src/data/reducer/editor.js [hunk 2]
```diff
@@ -36,9 +36,11 @@ import * as constants from '../../constants';
 const DEFAULT_STATE = {
     documents: [{
         contentType: constants.ContentType_BotChat,
+        title: "My bot",
```
Comment (eanders-MS) [inline client-v2/src/data/reducer/editor.js]: `title` should not appear here. The source of the title depends on the asset type. For cards, cardTab.js will have knowledge of how to read and display the card's title.

File: client-v2/src/index.tsx [hunk 3]
```diff
@@ -61,4 +61,4 @@ ReactDOM.render(
     document.getElementById('root')
 );
 
-registerServiceWorker();
+// registerServiceWorker();
```
Comment (eanders-MS) [inline client-v2/src/index.tsx]: Unrelated change?

File: client-v2/src/v1/mainView.tsx [hunk 4]
```diff
@@ -325,11 +325,11 @@ export class MainView extends React.Component<{}, {}> {
                                 onChange: (size: number) => this.verticalSplitChange(size)
                             },
                             [
-                                <div className='fill-parent'>
+                                <div key={ "botchat" } className='fill-parent'>
```
Comment (eanders-MS) [inline client-v2/src/v1/mainView.tsx]: Unrelated change?

File: .vscode/settings.json [hunk 5]
```diff
@@ -5,7 +5,7 @@
   "files.trimTrailingWhitespace": true,
   "files.exclude": {
     "client-v2/build": true,
-    "client-v2/public/external": true,
+    "client-v2/public/external": false,
```
Comment (eanders-MS) [inline .vscode/settings.json]: Unrelated change?

File: client-v2/src/data/action/cardActions.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+export const CARD_UPDATE_JSON = "CARD_UPDATE_JSON";
+export const CARD_ADD_OUTPUT_MSG = "CARD_ADD_OUTPUT_MSG";
+export const CARD_CLEAR_OUTPUT_WINDOW = "CARD_CLEAR_OUTPUT_WINDOW";
+
+export function updateCardJson(json) {
+    return {
+        type: CARD_UPDATE_JSON,
+        state: { json: json }
```
Comment (eanders-MS) [inline client-v2/src/data/action/cardActions.js]: Rename `state` to `payload` to make it a standard [FSA](https://github.com/acdlite/flux-standard-action). (True for each action here)

File: client-v2/src/data/action/cardActions.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
+export const CARD_UPDATE_JSON = "CARD_UPDATE_JSON";
```
Comment (eanders-MS) [inline client-v2/src/data/action/cardActions.js]: This file needs the copyright notice.

File: client-v2/src/data/reducer/card.js [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+const DEFAULT_STATE = {
+    title: "My card",
```
Comment (eanders-MS) [inline client-v2/src/data/reducer/card.js]: CardTab component should read it's tab label from this field (rather than from the document object in the store).

File: client-v2/src/index.tsx (hunk 3 above)
Comment (eanders-MS) [reply client-v2/src/index.tsx]: Probably a fine change to make though. @compulim thoughts?

File: client-v2/src/ui/editor/cardEditor/cardJsonEditor/index.js [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+import { connect } from 'react-redux'
+import { css } from 'glamor'
+import React from 'react'
+import ReactDOM from 'react-dom';
+import * as CardActions from '../../../../data/action/cardActions';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import { updateCardJson } from '../../../../data/action/cardActions';
+const fs = typeof window !== 'undefined' ? window.require('fs') : require('fs');
```
Comment (eanders-MS) [inline client-v2/src/ui/editor/cardEditor/cardJsonEditor/index.js]: I think a better pattern here is to break this into two lines:
```js
if (typeof window !== 'undefined') { require = window['require']; }
const fs = require('fs');
```
This way, it fixes `require` for all calls to it within the file.

File: client-v2/src/ui/editor/cardEditor/cardJsonEditor/index.js [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+const CSS = css({
```
Comment (eanders-MS) [inline client-v2/src/ui/editor/cardEditor/cardJsonEditor/index.js]: File needs copyright header.

File: client-v2/src/v1/mainView.tsx (hunk 4 above)
Comment (eanders-MS) [reply client-v2/src/v1/mainView.tsx]: (probably fine.. it's all getting reimplemented anyway)

File: client-v2/src/index.tsx (hunk 3 above)
Comment (Fignewton5) [reply client-v2/src/index.tsx]: This was suggested by @compulim 👍

File: client-v2/src/v1/mainView.tsx (hunk 4 above)
Comment (Fignewton5) [reply client-v2/src/v1/mainView.tsx]: These were just to suppress the annoying React warnings for arrayed elements being rendered without unique keys. I can remove if necessary

File: client-v2/src/data/action/cardActions.js (hunk 6 above)
Comment (Fignewton5) [reply client-v2/src/data/action/cardActions.js]: Was completely unaware of this standard. Good call 👍

File: client-v2/src/data/reducer/editor.js (hunk 2 above)
Comment (Fignewton5) [reply client-v2/src/data/reducer/editor.js]: Ok, so will the asset type's structure populate the `content` object? I'm just trying to think about how the `cardExplorer.js` will know about the card's title. Will it look for the title in `document.content.title` in the case of a card?

File: client-v2/src/ui/shell/explorer/cardExplorer.js [hunk 11]
```diff
@@ -54,19 +55,34 @@ const BOTS_CSS = css({
...
+                        </ExpandCollapseContent>
+                    </ExpandCollapse>
+                </li>
+            </ul>
+        );
+    }
+}
+
+export default connect(state => ({
+    cards: state.editor.documents.filter(doc => doc.contentType === constants.ContentType_Card)
```
Comment (eanders-MS) [inline client-v2/src/ui/shell/explorer/cardExplorer.js]: Issue identified by @compulim:
> This call to Array.filter() will create a new documents array every time connect is called. Two things to do, either:
> 1.	Just pass documents inside the React component and process it there (in render or in willReceiveProps -> setState)
> 2.	Use reselect or any memorize function, n=1 memoizer is okay

I chose to just pass the documents, but not sure it isn't still making a copy in my render function.
